File: innerlibs1/Printer/EscBema.vb
Imports System.Text
Imports InnerLibs.Printer
Imports InnerLibs.Printer.Command

Namespace EscBemaCommands

    Public Class EscBema
        Implements IPrintCommand

#Region "Properties"

        Public ReadOnly Property ColsNomal As Integer Implements IPrintCommand.ColsNomal
            Get
                Return 50
            End Get
        End Property

        Public ReadOnly Property ColsCondensed As Integer Implements IPrintCommand.ColsCondensed
            Get
                Return 67
            End Get
        End Property

        Public ReadOnly Property ColsExpanded As Integer Implements IPrintCommand.ColsExpanded
            Get
                Return 25
            End Get
        End Property

        Public Property Encoding As Encoding = Encoding.GetEncoding(850) Implements IPrintCommand.Encoding

#End Region



#Region "Methods"



        Public Function AutoTest() As Byte() Implements IPrintCommand.AutoTest
            Return New Byte() {&H1D, &HF9, &H29, &H30}
        End Function



        Private Shared Function StoreQr(ByVal qrData As String, ByVal size As QrCodeSize) As IEnumerable(Of Byte)
            Dim length = qrData.Length
            Dim b = CByte(length Mod 256)
            Dim b2 = CByte(length / 256)
            Return New Byte() {29, 107, 81}.AddBytes({size.ToByte()}).AddBytes((New Byte() {6, 0, 1})).AddBytes({b}).AddBytes({b2})
        End Function

        Public Function PrintQrData(ByVal qrData As String) As Byte() Implements IPrintCommand.PrintQrData
            Return PrintQrData(qrData, QrCodeSize.Size0)
        End Function

        Public Function PrintQrData(ByVal qrData As String, ByVal qrCodeSize As QrCodeSize) As Byte() Implements IPrintCommand.PrintQrData
            Dim list = New List(Of Byte)()
            list.AddRange(StoreQr(qrData, qrCodeSize))
            list.AddRange(Encoding.GetBytes(qrData))
            Return list.ToArray()
        End Function

        Public Function FullCut() As Byte() Implements IPrintCommand.FullCut
            Return New Byte() {27, "w"c.ToByte()}
        End Function

        Public Function PartialCut() As Byte() Implements IPrintCommand.PartialCut
            Return New Byte() {27, "m"c.ToByte()}
        End Function

        Public Function Initialize() As Byte() Implements IPrintCommand.Initialize
            Return New Byte() {27, "@"c.ToByte()}.AddBytes(SetEscBema()).AddBytes(SetLineSpace3())
        End Function

        Private Shared Function SetEscBema() As Byte()
            Return New Byte() {29, 249, 32, 0}
        End Function

        Private Shared Function SetLineSpace3(ByVal Optional range As Byte = 20) As Byte()
            Return New Byte() {27, 51, range}
        End Function

        Public Function NormalFont() As Byte() Implements IPrintCommand.NormalFont
            Return New Byte() {27, "W"c.ToByte(), 0, 27, "d"c.ToByte(), 0}
        End Function

        Public Function LargeFont() As Byte() Implements IPrintCommand.LargeFont
            Return New Byte() {27, "W"c.ToByte(), 1, 27, "d"c.ToByte(), 1}
        End Function

        Public Function LargerFont() As Byte() Implements IPrintCommand.LargerFont
            Return New Byte() {29, "!"c.ToByte(), 32}
        End Function



        Public Function Italic(ByVal state As Boolean) As Byte() Implements IPrintCommand.Italic
            Return If(state = True, New Byte() {27, "4"c.ToByte()}, New Byte() {27, "5"c.ToByte()})
        End Function



        Public Function Bold(ByVal state As Boolean) As Byte() Implements IPrintCommand.Bold
            Return If(state = True, New Byte() {27, "E"c.ToByte()}, New Byte() {27, "F"c.ToByte()})
        End Function



        Public Function Underline(ByVal state As Boolean) As Byte() Implements IPrintCommand.Underline
            Return If(state = True, New Byte() {27, "-"c.ToByte(), 1}, New Byte() {27, "-"c.ToByte(), 0})
        End Function



        Public Function Expanded(ByVal state As Boolean) As Byte() Implements IPrintCommand.Expanded
            Return New Byte() {27, "W"c.ToByte(), If(state = True, "1"c.ToByte(), "0"c.ToByte())}
        End Function



        Public Function Condensed(ByVal state As Boolean) As Byte() Implements IPrintCommand.Condensed
            Return If(state = True, New Byte() {27, 15}, New Byte() {27, "P"c.ToByte()})
        End Function

        Public Function OpenDrawer() As Byte() Implements IPrintCommand.OpenDrawer
            Return New Byte() {27, "v"c.ToByte(), 140}
        End Function

        Public Function Code128(ByVal code As String) As Byte() Implements IPrintCommand.Code128
            Return New Byte() {29, 119, 2}.AddBytes((New Byte() {29, 104, 50})).AddBytes((New Byte() {29, 102, 1})).AddBytes((New Byte() {29, 72, 0})).AddBytes((New Byte() {29, 107, 73})).AddBytes({CByte(code.Length + 2)}).AddBytes({"{"c.ToByte, "C"c.ToByte}).AddTextBytes(code, Encoding) ' Width
            ' Height
            ' font hri character
            ' If print code informed
            ' printCode
        End Function

        Public Function Code39(ByVal code As String) As Byte() Implements IPrintCommand.Code39
            Return New Byte() {29, 119, 2}.AddBytes((New Byte() {29, 104, 50})).AddBytes((New Byte() {29, 102, 0})).AddBytes((New Byte() {29, 72, 0})).AddBytes((New Byte() {29, 107, 4})).AddTextBytes(code, Encoding).AddBytes(New Byte() {0}) ' Width
            ' Height
            ' font hri character
            ' If print code informed
        End Function

        Public Function Ean13(ByVal code As String) As Byte() Implements IPrintCommand.Ean13
            If code.Trim().Length <> 13 Then Return New Byte(-1) {}
            Return New Byte() {29, 119, 40}.AddBytes((New Byte() {29, 104, 50})).AddBytes((New Byte() {29, 72, 0})).AddBytes((New Byte() {29, 107, 67, 12})).AddTextBytes(code.Substring(0, 12), Encoding) ' Width
            ' Height
            ' If print code informed
        End Function

        Private Shared Function Align(ByVal justification As Justifications) As Byte()
            Dim lAlign As Byte

            Select Case justification
                Case Justifications.Right
                    lAlign = "2"c.ToByte()
                Case Justifications.Center
                    lAlign = "1"c.ToByte()
                Case Else
                    lAlign = "0"c.ToByte()
            End Select

            Return New Byte() {27, "a"c.ToByte(), lAlign}
        End Function

        Public Function Left() As Byte() Implements IPrintCommand.Left
            Return Align(Justifications.Left)
        End Function

        Public Function Right() As Byte() Implements IPrintCommand.Right
            Return Align(Justifications.Right)
        End Function

        Public Function Center() As Byte() Implements IPrintCommand.Center
            Return Align(Justifications.Center)
        End Function

        Public Function PrintImage(image As Drawing.Image, highDensity As Boolean) As Byte() Implements IPrintCommand.PrintImage
            Return SharedImagePrinter(image, highDensity)
        End Function

#End Region

    End Class

End Namespace

File: InnerLibs/Printer/EscBema.cs
using System;
using System.Collections.Generic;
using System.Text;
using InnerLibs.Printer;
using InnerLibs.Printer.Command;

namespace InnerLibs.EscBemaCommands
{
    public class EscBema : IPrintCommand
    {

        #region Properties

        public int ColsNomal
        {
            get
            {
                return 50;
            }
        }

        public int ColsCondensed
        {
            get
            {
                return 67;
            }
        }

        public int ColsExpanded
        {
            get
            {
                return 25;
            }
        }

        public Encoding Encoding { get; set; } = Encoding.GetEncoding(850);

        #endregion



        #region Methods



        public byte[] AutoTest()
        {
            return new byte[] { 0x1D, 0xF9, 0x29, 0x30 };
        }

        private static IEnumerable<byte> StoreQr(string qrData, QrCodeSize size)
        {
            int length = qrData.Length;
            byte b = (byte)(length % 256);
            byte b2 = (byte)Math.Round(length / 256d);
            return (new byte[] { 29, 107, 81 }).AddBytes(new[] { size.ToByte() }).AddBytes(new byte[] { 6, 0, 1 }).AddBytes(new[] { b }).AddBytes(new[] { b2 });
        }

        public byte[] PrintQrData(string qrData)
        {
            return PrintQrData(qrData, QrCodeSize.Size0);
        }

        public byte[] PrintQrData(string qrData, QrCodeSize qrCodeSize)
        {
            var list = new List<byte>();
            list.AddRange(StoreQr(qrData, qrCodeSize));
            list.AddRange(Encoding.GetBytes(qrData));
            return list.ToArray();
        }

        public byte[] FullCut()
        {
            return new byte[] { 27, 'w'.ToByte() };
        }

        public byte[] PartialCut()
        {
            return new byte[] { 27, 'm'.ToByte() };
        }

        public byte[] Initialize()
        {
            return (new byte[] { 27, '@'.ToByte() }).AddBytes(SetEscBema()).AddBytes(SetLineSpace3());
        }

        private static byte[] SetEscBema()
        {
            return new byte[] { 29, 249, 32, 0 };
        }

        private static byte[] SetLineSpace3(byte range = 20)
        {
            return new byte[] { 27, 51, range };
        }

        public byte[] NormalFont()
        {
            return new byte[] { 27, 'W'.ToByte(), 0, 27, 'd'.ToByte(), 0 };
        }

        public byte[] LargeFont()
        {
            return new byte[] { 27, 'W'.ToByte(), 1, 27, 'd'.ToByte(), 1 };
        }

        public byte[] LargerFont()
        {
            return new byte[] { 29, '!'.ToByte(), 32 };
        }

        public byte[] Italic(bool state)
        {
            return state == true ? (new byte[] { 27, '4'.ToByte() }) : (new byte[] { 27, '5'.ToByte() });
        }

        public byte[] Bold(bool state)
        {
            return state == true ? (new byte[] { 27, 'E'.ToByte() }) : (new byte[] { 27, 'F'.ToByte() });
        }

        public byte[] Underline(bool state)
        {
            return state == true ? (new byte[] { 27, '-'.ToByte(), 1 }) : (new byte[] { 27, '-'.ToByte(), 0 });
        }

        public byte[] Expanded(bool state)
        {
            return new byte[] { 27, 'W'.ToByte(), state == true ? '1'.ToByte() : '0'.ToByte() };
        }

        public byte[] Condensed(bool state)
        {
            return state == true ? (new byte[] { 27, 15 }) : (new byte[] { 27, 'P'.ToByte() });
        }

        public byte[] OpenDrawer()
        {
            return new byte[] { 27, 'v'.ToByte(), 140 };
        }

        public byte[] Code128(string code)
        {
            return (new byte[] { 29, 119, 2 }).AddBytes(new byte[] { 29, 104, 50 }).AddBytes(new byte[] { 29, 102, 1 }).AddBytes(new byte[] { 29, 72, 0 }).AddBytes(new byte[] { 29, 107, 73 }).AddBytes(new[] { (byte)(code.Length + 2) }).AddBytes(new[] { '{'.ToByte(), 'C'.ToByte() }).AddTextBytes(code, Encoding); // Width
            // Height
            // font hri character
            // If print code informed
            // printCode
        }

        public byte[] Code39(string code)
        {
            return (new byte[] { 29, 119, 2 }).AddBytes(new byte[] { 29, 104, 50 }).AddBytes(new byte[] { 29, 102, 0 }).AddBytes(new byte[] { 29, 72, 0 }).AddBytes(new byte[] { 29, 107, 4 }).AddTextBytes(code, Encoding).AddBytes(new byte[] { 0 }); // Width
            // Height
            // font hri character
            // If print code informed
        }

        public byte[] Ean13(string code)
        {
            if (code.Trim().Length != 13)
                return new byte[0];
            return (new byte[] { 29, 119, 40 }).AddBytes(new byte[] { 29, 104, 50 }).AddBytes(new byte[] { 29, 72, 0 }).AddBytes(new byte[] { 29, 107, 67, 12 }).AddTextBytes(code.Substring(0, 12), Encoding); // Width
            // Height
            // If print code informed
        }

        private static byte[] Align(Justifications justification)
        {
            byte lAlign;
            switch (justification)
            {
                case Justifications.Right:
                    {
                        lAlign = '2'.ToByte();
                        break;
                    }

                case Justifications.Center:
                    {
                        lAlign = '1'.ToByte();
                        break;
                    }

                default:
                    {
                        lAlign = '0'.ToByte();
                        break;
                    }
            }

            return new byte[] { 27, 'a'.ToByte(), lAlign };
        }

        public byte[] Left()
        {
            return Align(Justifications.Left);
        }

        public byte[] Right()
        {
            return Align(Justifications.Right);
        }

        public byte[] Center()
        {
            return Align(Justifications.Center);
        }

        public byte[] PrintImage(System.Drawing.Image image, bool highDensity)
        {
            return PrinterByteExtensions.SharedImagePrinter(image, highDensity);
        }

        #endregion

    }
}